### PR TITLE
 Suggest user merge if user with same email address exists 

### LINKED ISSUE #1958: Suggest user merge if email address already exists
As a manager, when changing a user's email address to an existing email address of another user, an error is shown in the user form.
In addition, a warning message should be shown on top of the page, suggesting to merge the two users with a link to the merge page (where the user currently edited will be the main user).

### PR DESCRIPTION
**Pull request by hansegucker**
Close #1958

### CODE REVIEW

File: evap/staff/forms.py [hunk 1]
```diff
@@ -999,6 +1000,7 @@ def clean_email(self):
             user_with_same_email = user_with_same_email.exclude(pk=self.instance.pk)
 
         if user_with_same_email.exists():
+            self.user_with_same_email = user_with_same_email.get()
```
Comment (niklasmohrin) [inline evap/staff/forms.py:1003]: This now does two database queries, one for `exists` and one for `get`. In general, could it be the case that multiple users exist? @janno42

File: evap/staff/templates/staff_user_form.html [hunk 2]
```diff
@@ -32,6 +32,16 @@ <h3>{% if form.instance.id %}{% trans 'Edit user' %}{% else %}{% trans 'Create u
         {% endif %}
     </div>
 
+    {% if user_with_same_email %}
+        <div class="alert alert-warning alert-dismissible">
+            <p>
+                {% trans "There is already another user with this email address. You probably want to merge the users." %}
+            </p>
+            <a type="button" class="btn btn-primary btn-sm" href="{% url 'staff:user_merge' form.instance.id user_with_same_email.id %}">{% trans "Merge both users" %}</a>
```
Comment (janno42) [inline evap/staff/templates/staff_user_form.html:40]:
```suggestion
            <a type="button" class="btn btn-primary btn-sm" href="{% url 'staff:user_merge' user_with_same_email.id form.instance.id %}">{% trans "Merge both users" %}</a>
```

In the specification, I wrote it the wrong way round - please make the existing user the main user of the merge.

File: evap/staff/forms.py (hunk 1 above)
Comment (hansegucker) [reply evap/staff/forms.py:1003]: No, because `email` is unique on database level.

File: evap/staff/templates/staff_user_form.html [hunk 3]
```diff
@@ -32,6 +32,16 @@ <h3>{% if form.instance.id %}{% trans 'Edit user' %}{% else %}{% trans 'Create u
         {% endif %}
     </div>
 
+    {% if user_with_same_email %}
+        <div class="alert alert-warning alert-dismissible">
+            <p>
+                {% trans "There is already another user with this email address. You probably want to merge the users." %}
```
Comment (janno42) [inline evap/staff/templates/staff_user_form.html:38]:
```suggestion
                {% trans "A user with this email address already exists. You probably want to merge the users." %}
```

**Review by janno42 — CHANGES_REQUESTED** [review]

**Review by janno42 — APPROVED** [review]
✔️ Meets requirements
✔️ UI functionality checked

File: evap/staff/forms.py (hunk 1 above)
Comment (niklasmohrin) [reply evap/staff/forms.py:1003]: Except that we are matching with `iexact` above. I think we can usually assume that all emails in the database are lowercased, but I wonder if we should handle the other case here (or assert that at most one matches?)

Comment (richardebeling) [reply evap/staff/forms.py:1003]: I think it's fair if we just take the first email that matches. In the (unlikely) case there are multiple matching emails, you'd simply get another merge suggestion afterwards.

File: evap/staff/tests/test_views.py [hunk 4]
```diff
@@ -272,13 +272,25 @@ def setUpTestData(cls):
...
         page = self.app.get(self.url, user=self.manager, status=200)
         form = page.forms["user-form"]
         form["email"] = "lfo9e7bmxp1xi@institution.example.com"
         form.submit()
         self.assertTrue(UserProfile.objects.filter(email="lfo9e7bmxp1xi@institution.example.com").exists())
 
+    def test_user_edit_duplicate_email(self):
+        second_user = baker.make(UserProfile)
+        second_user.email = "lfo9e7bmxp1xi@institution.example.com"
+        second_user.save()
```
Comment (richardebeling) [inline evap/staff/tests/test_views.py:285]:
```suggestion
        second_user = baker.make(UserProfile, email="lfo9e7bmxp1xi@institution.example.com")
```

File: evap/staff/tests/test_views.py [hunk 5]
```diff
@@ -272,13 +272,25 @@ def setUpTestData(cls):
...
+        page = self.app.get(self.url, user=self.manager, status=200)
+        form = page.forms["user-form"]
+        form["email"] = "lfo9e7bmxp1xi@institution.example.com"
```
Comment (richardebeling) [inline evap/staff/tests/test_views.py:288]: If you reuse the value, I as a reader don't have to spend mental energy on comparing the value:
```suggestion
        form["email"] = second_user.email
```

File: evap/staff/tests/test_views.py (hunk 4 above)
Comment (richardebeling) [reply evap/staff/tests/test_views.py:285]: (we could also use a less cryptic looking mailbox-part than "lfo9e7bmxp1xi" here. I don't know why the existing test used that, may have been about accidental collisions, but that shouldn't be a problem anymore)

File: evap/staff/forms.py (hunk 1 above)
Comment (hansegucker) [reply evap/staff/forms.py]: Done

**Comment by hansegucker** [comment]
@richardebeling Please review again.

File: evap/staff/tests/test_views.py [hunk 6]
```diff
@@ -272,12 +272,24 @@ def setUpTestData(cls):
...
-        form["email"] = "lfo9e7bmxp1xi@institution.example.com"
+        form["email"] = "test@institution.example.com"
         form.submit()
-        self.assertTrue(UserProfile.objects.filter(email="lfo9e7bmxp1xi@institution.example.com").exists())
+        self.assertTrue(UserProfile.objects.filter(email="test@institution.example.com").exists())
+
+    def test_user_edit_duplicate_email(self):
+        second_user = baker.make(UserProfile)
+        second_user.email = "test@institution.example.com"
+        second_user.save()
```
Comment (richardebeling) [inline evap/staff/tests/test_views.py:285]:
```suggestion
        second_user = baker.make(UserProfile, email="test@institution.example.com")
```

**Review by richardebeling — APPROVED** [review]

**Review by niklasmohrin — APPROVED** [review]
(lgtm, once the other comment is fixed)

**Comment by richardebeling** [comment]
thanks :)

**Pull request merged**